File: js/test_rest.js
/**
 * Test script for marketdata-js REST client
 *
 * This script tests:
 * 1. Module loading
 * 2. RestClient creation
 * 3. Client chain access (stock/futopt -> intraday)
 * 4. Async/Promise API behavior
 * 5. Error handling with error codes
 */

const { RestClient } = require('./index.js');

console.log('=== marketdata-js REST Client Tests ===\n');

// Test 1: Module loading
console.log('Test 1: Module loading');
try {
    console.log('  - RestClient class:', typeof RestClient === 'function' ? 'OK' : 'FAIL');
} catch (e) {
    console.log('  - FAIL:', e.message);
}

// Test 2: Client creation
console.log('\nTest 2: Client creation');
try {
    const client = new RestClient('test-api-key');
    console.log('  - new RestClient():', client ? 'OK' : 'FAIL');
} catch (e) {
    console.log('  - FAIL:', e.message);
}

// Test 3: Client chain access
console.log('\nTest 3: Client chain access');
try {
    const client = new RestClient('test-api-key');

    // Stock chain
    const stock = client.stock;
    console.log('  - client.stock:', typeof stock === 'object' ? 'OK' : 'FAIL');

    const stockIntraday = stock.intraday;
    console.log('  - client.stock.intraday:', typeof stockIntraday === 'object' ? 'OK' : 'FAIL');

    // FutOpt chain
    const futopt = client.futopt;
    console.log('  - client.futopt:', typeof futopt === 'object' ? 'OK' : 'FAIL');

    const futoptIntraday = futopt.intraday;
    console.log('  - client.futopt.intraday:', typeof futoptIntraday === 'object' ? 'OK' : 'FAIL');

    // Method existence
    console.log('  - quote method:', typeof stockIntraday.quote === 'function' ? 'OK' : 'FAIL');
    console.log('  - ticker method:', typeof stockIntraday.ticker === 'function' ? 'OK' : 'FAIL');
    console.log('  - candles method:', typeof stockIntraday.candles === 'function' ? 'OK' : 'FAIL');
    console.log('  - trades method:', typeof stockIntraday.trades === 'function' ? 'OK' : 'FAIL');
    console.log('  - volumes method:', typeof stockIntraday.volumes === 'function' ? 'OK' : 'FAIL');
    console.log('  - futopt.products method:', typeof futoptIntraday.products === 'function' ? 'OK' : 'FAIL');
} catch (e) {
    console.log('  - FAIL:', e.message);
}

// Test 4: Async/Promise API
console.log('\nTest 4: Async/Promise API');
(async () => {
    try {
        const client = new RestClient('test-api-key');

        // Test stock methods return Promises
        const stockQuote = client.stock.intraday.quote('2330');
        const stockTicker = client.stock.intraday.ticker('2330');
        const stockCandles = client.stock.intraday.candles('2330', '5');
        const stockTrades = client.stock.intraday.trades('2330');
        const stockVolumes = client.stock.intraday.volumes('2330');

        console.log('  - stock.quote() returns Promise:', stockQuote instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - stock.ticker() returns Promise:', stockTicker instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - stock.candles() returns Promise:', stockCandles instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - stock.trades() returns Promise:', stockTrades instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - stock.volumes() returns Promise:', stockVolumes instanceof Promise ? 'OK' : 'FAIL');

        // Suppress rejections
        stockQuote.catch(() => {});
        stockTicker.catch(() => {});
        stockCandles.catch(() => {});
        stockTrades.catch(() => {});
        stockVolumes.catch(() => {});

        // Test futopt methods return Promises
        const futoptQuote = client.futopt.intraday.quote('TXFC4');
        const futoptTicker = client.futopt.intraday.ticker('TXFC4');
        const futoptCandles = client.futopt.intraday.candles('TXFC4', '5');
        const futoptTrades = client.futopt.intraday.trades('TXFC4');
        const futoptVolumes = client.futopt.intraday.volumes('TXFC4');
        const futoptProducts = client.futopt.intraday.products('FUTURE');

        console.log('  - futopt.quote() returns Promise:', futoptQuote instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - futopt.ticker() returns Promise:', futoptTicker instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - futopt.candles() returns Promise:', futoptCandles instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - futopt.trades() returns Promise:', futoptTrades instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - futopt.volumes() returns Promise:', futoptVolumes instanceof Promise ? 'OK' : 'FAIL');
        console.log('  - futopt.products() returns Promise:', futoptProducts instanceof Promise ? 'OK' : 'FAIL');

        // Suppress rejections
        futoptQuote.catch(() => {});
        futoptTicker.catch(() => {});
        futoptCandles.catch(() => {});
        futoptTrades.catch(() => {});
        futoptVolumes.catch(() => {});
        futoptProducts.catch(() => {});

        // Test 5: Error handling with error codes
        console.log('\nTest 5: Error handling');

        // Test products with invalid type - should return validation error synchronously
        try {
            const result = await client.futopt.intraday.products('INVALID_TYPE');
            console.log('  - Invalid type validation: Unexpected success');
        } catch (err) {
            const isValidationError = err.message.includes('FUTURE') || err.message.includes('OPTION');
            console.log('  - Invalid type validation:', isValidationError ? 'OK' : 'FAIL');
            console.log('    Error:', err.message);
        }

        // API call will fail with invalid API key
        try {
            await client.stock.intraday.quote('2330');
            console.log('  - Auth error handling: Unexpected success');
        } catch (err) {
            const hasErrorCode = err.message.match(/\[\d+\]/);
            console.log('  - Error thrown properly:', err instanceof Error ? 'OK' : 'FAIL');
            console.log('  - Error has code format [XXXX]:', hasErrorCode ? 'OK' : 'FAIL');
            console.log('    Error:', err.message.substring(0, 80));
        }

        // Test 6: TypeScript types check (index.d.ts exists)
        console.log('\nTest 6: TypeScript definitions');
        const fs = require('fs');
        const path = require('path');
        const dtsPath = path.join(__dirname, 'index.d.ts');
        const exists = fs.existsSync(dtsPath);
        console.log('  - index.d.ts exists:', exists ? 'OK' : 'FAIL');

        if (exists) {
            const content = fs.readFileSync(dtsPath, 'utf8');
            console.log('  - Contains RestClient:', content.includes('class RestClient') ? 'OK' : 'FAIL');
            console.log('  - Contains StockClient:', content.includes('class StockClient') ? 'OK' : 'FAIL');
            console.log('  - Contains FutOptClient:', content.includes('class FutOptClient') ? 'OK' : 'FAIL');
            console.log('  - Contains async quote method:', content.includes('quote(symbol: string): Promise') ? 'OK' : 'FAIL');
        }

        console.log('\n=== Tests Complete ===');
    } catch (e) {
        console.log('  - Test setup FAIL:', e.message);
    }
})();

File: py/test_rest.py
"""
Test script for fugle_marketdata REST client

This script tests:
1. Module import
2. RestClient creation with different auth methods
3. Client chain access (client.stock.intraday, client.futopt.intraday)
4. Error handling with error_code verification
"""

import os
import fugle_marketdata
from fugle_marketdata import RestClient, MarketDataError


def test_import():
    """Test module can be imported"""
    print("Testing import...")
    assert hasattr(fugle_marketdata, 'RestClient')
    assert hasattr(fugle_marketdata, 'MarketDataError')
    assert hasattr(fugle_marketdata, 'StockClient')
    assert hasattr(fugle_marketdata, 'StockIntradayClient')
    assert hasattr(fugle_marketdata, 'FutOptClient')
    assert hasattr(fugle_marketdata, 'FutOptIntradayClient')
    print("  Import OK")


def test_client_creation():
    """Test RestClient creation with different auth methods"""
    print("Testing client creation...")

    # API key authentication
    client1 = RestClient("test-api-key")
    print(f"  API key client: {client1}")

    # Bearer token authentication
    client2 = RestClient.with_bearer_token("test-token")
    print(f"  Bearer token client: {client2}")

    # SDK token authentication
    client3 = RestClient.with_sdk_token("test-sdk-token")
    print(f"  SDK token client: {client3}")

    print("  Client creation OK")


def test_client_chain():
    """Test client chain access pattern"""
    print("Testing client chain...")

    # 從環境變數取得 API key，若無則用測試用 key
    api_key = os.environ.get("FUGLE_API_KEY", "test-api-key")
    client = RestClient(api_key)

    # Stock chain
    stock = client.stock
    assert stock is not None
    print(f"  client.stock: {stock}")

    stock_intraday = client.stock.intraday
    assert stock_intraday is not None
    print(f"  client.stock.intraday: {stock_intraday}")

    # FutOpt chain
    futopt = client.futopt
    assert futopt is not None
    print(f"  client.futopt: {futopt}")

    futopt_intraday = client.futopt.intraday
    assert futopt_intraday is not None
    print(f"  client.futopt.intraday: {futopt_intraday}")

    print("  Client chain OK")


def test_error_handling():
    """Test error handling with error_code"""
    print("Testing error handling...")

    # MarketDataError should be an exception type
    assert issubclass(MarketDataError, Exception)
    print(f"  MarketDataError is Exception subclass: True")

    # Test that MarketDataError can be raised and caught
    try:
        raise MarketDataError("Test error", 1001)
    except MarketDataError as e:
        assert str(e.args[0]) == "Test error"
        assert e.args[1] == 1001  # error_code in args[1]
        print(f"  Error message: {e.args[0]}")
        print(f"  Error code: {e.args[1]}")

    print("  Error handling OK")


def test_api_call_with_invalid_key():
    """Test API call with invalid key returns proper error"""
    print("Testing API call with invalid key...")

    client = RestClient("invalid-key-that-will-fail")

    try:
        # This will fail because the API key is invalid
        # We expect an authentication or API error
        quote = client.stock.intraday.quote("2330")
        print(f"  Warning: API call succeeded unexpectedly: {quote}")
    except MarketDataError as e:
        message = e.args[0]
        error_code = e.args[1]
        print(f"  Expected error caught!")
        print(f"  Error message: {message}")
        print(f"  Error code: {error_code}")
        # Error codes:
        # 2002 = AuthError
        # 2003 = ApiError
        # 2001 = ConnectionError
        assert error_code in [2001, 2002, 2003], f"Unexpected error code: {error_code}"
    except Exception as e:
        print(f"  Unexpected exception type: {type(e).__name__}: {e}")
        raise

    print("  API call error handling OK")


def run_all_tests():
    """Run all tests"""
    print("=" * 60)
    print("fugle_marketdata REST Client Tests")
    print("=" * 60)

    test_import()
    test_client_creation()
    test_client_chain()
    test_error_handling()
    test_api_call_with_invalid_key()

    print("=" * 60)
    print("All tests passed!")
    print("=" * 60)


if __name__ == "__main__":
    run_all_tests()
